Add metasources to l10nregistry

dminor opened the pull request:
This adds metasource to FileSource and groups sources by metasource in the registry. Only sources with the same metasource are passed into the solver.

This is WIP, so far it is only implemented for the synchronous registry. This adds one test, I've checked manually using it that the bundle is built from the "langpack" sources as expected. Next steps are support for the asynchronous registry and further tests.

Fixes #12.

dminor committed:
Add benchmark dependencies

dminor committed:
Fix clippy complex type warning

dminor committed:
Add extra documentation for FileSource

dminor committed:
Add metasources and use them for synchronous registry

This adds a metasource to FileSource. The metasource is used to group related
sources together. Only sources from the same metasource will be passed to
the solver to find solutions.

This updates the synchronous registry to use metasources. Updates for the
asychronous registry will follow in the next commit.

zbraniecki commented:
I took a look at the code and it looks good!

One suggestion is to add a benchmark that would show a scenario where currently we do catastrophic and with your PR we'd fix it.

The PR #11 does something like that but the testing scenario against benchmark `solver` doesn't recognize metasource. You could either improve that so that we can describe scenarios that without your PR are "single metasource" and with your PR can be separated into two metasources, or come up with some separate patch on higher level than solver.

dminor committed:
Add documentation for Scenario

dminor committed:
Add registry benchmark

Because metasources are handled at the registry level and not at the solver
level, we need to add a new benchmark to see the effect of using metasources.

dminor committed:
Refactor metasource methods

dminor committed:
Add metasources support to asynchronous registry

This also fixes a problem where only checked the first metasource
was checked in the synchronous registry.

dminor committed:
Add parallel benchmark

dminor commented:
Initial benchmark results:
```
non-metasource/serial   time:   [11.079 s 11.114 s 11.156 s]                                   
                        change: [-0.5781% -0.0475% +0.4713%] (p = 0.87 > 0.05)
                        No change in performance detected.
Found 2 outliers among 10 measurements (20.00%)
  2 (20.00%) high mild
non-metasource/parallel time:   [21.920 ms 21.956 ms 21.981 ms]                                   
                        change: [-1.8610% -1.5773% -1.3213%] (p = 0.00 < 0.05)
                        Performance has improved.

metasource/serial       time:   [214.16 us 214.57 us 215.44 us]                             
                        change: [-1.0770% -0.6027% -0.0807%] (p = 0.04 < 0.05)
                        Change within noise threshold.
Found 1 outliers among 10 measurements (10.00%)
  1 (10.00%) high mild
metasource/parallel     time:   [228.90 us 229.33 us 229.72 us]                               
                        change: [+0.0771% +0.2694% +0.4547%] (p = 0.02 < 0.05)
                        Change within noise threshold.
```

dminor committed:
Bump version to 0.3.0

dminor committed:
Run cargo fmt

dminor committed:
Add FileSource::new_with_index_and_metasource

dminor committed:
Handle empty registry properly

dminor committed:
Bug fixes

This adds code to try the next metasource if we're unable to locate
a resource file. This is a normal occurence with langpacks for
multiple languages installed.

This also fixes a bug where we would keep an empty metasource around
after all of the sources were removed from it.

Finally, it adds metasource to the Debug implementation for FileSource.

gregtatum commented:
Adding myself as a non-blocking reviewer.

zbraniecki reviewed (approved):
this looks great!

The only suggestion is to move all constructors to take metasource as `Option`.

gregtatum commented on src/source/mod.rs:
```diff
@@ -64,8 +64,13 @@ impl Future for ResourceStatus {
     pub name: String,
+    /// Pre-formatted path for the FileSource, e.g. "/browser/data/locale/{locale}/"
     pub pre_path: String,
+    /// Meta-source name for the FileSource, e.g. "browser", "lang-pack"
```
Suggestion (docs): I don't find this answered in the docs:

* What exactly is a meta source?
* Why does it exist?
* How does a "browser" or "lang-pack" map to an actual location?

dminor committed:
Move metasource to option in constructors

zbraniecki commented:
lgtm!

gregtatum reviewed (commented):
Don't block on me, I'm still going to go through everything more so I can understand it, but here is my one comment so far.
